### ISSUE TITLE
libripgrep example

### ISSUE THREAD

**Issue by jessegrosjean** [BurntSushi/ripgrep#1009]
#### What version of ripgrep are you using?

https://github.com/BurntSushi/ripgrep/tree/ag/libripgrep-freeze-2

#### Describe your question, feature request, or bug.

I'm excited to see the [libripgrep-freeze-2 work](https://github.com/BurntSushi/ripgrep/tree/ag/libripgrep-freeze-2). I'd love to see a best practices example that recreates `rg --count` with the lib.

I've created a starting point here:

https://github.com/jessegrosjean/libripgrep_example

But I'm new to rust as of yesterday, and don't have a good idea of what I'm doing. I hope someone can take a look at that code and let me know how to improve it. And eventually I hope same/similar code will get included as an example with libripgrep.

**Comment by BurntSushi** [comment]
I'm excited that you are already trying to jump into libripgrep, but it isn't on master yet and thus still in development. I created those branches for folks to try out some new ripgrep features but don't really intend it to be anything more than that.

It is indeed the case that a major part of the remaining work is doumentation and examples, among other things, such as additional integration tests to cover new features. There is already a fair bit of docs, but there is basically no high level documentation.

I will take a look at your code and give feedback when I get a chance, but I don't want to get too caught up with thia before I'm actually ready to publish libripgrep.

**Comment by BurntSushi** [comment]
@jessegrosjean Sorry I didn't add this in my previous comment, but I'm super impressed you were able to get a working program while also being simultaneously new to Rust and working without any high level docs or examples. I've toiled quite a bit over the various APIs and worry quite a bit whether something is more complex than it needs to be. Abstraction has steep costs...

**Comment by jessegrosjean** [comment]
I'm just headed away from computer, but looking forward to digging through all this in a day or two when I get back. Thanks so much!

**Comment by jessegrosjean** [comment]
> This is pretty similar to the first revision, but I removed your custom implementation of Sink (nice work on figuring that out!) and replaced it with a built-in implementation of Sink that happens to give you what you need here.

Bytes simplifies things a lot.

I actually started with that approach, but decided I was in the wrong place because I was expecting some solution that would give me a list of matching ranges in a callback. Then I started digging into the other Sink implementations since I knew they generated ranges somehow. And that's where I eventually learned that I would need to run the matcher over the given content to get the ranges... but by that time I'd forgotten that I was already in that position when I started with Bytes.

Makes sense now, I wouldn't want to have to pay the cost of iterating over those matches if I didn't need them. Helper `submatches()` function on Sink would have allowed me to figure things out easier?

>  your program run literally as fast as ripgrep does on the Linux kernel repo.

Yeah, I love this! And actually it runs a bit faster since I can call it through a C API directly and don't need to worry about firing up a subprocess and parsing text results. Also I get to add filename matching logic to the same pass instead of having to perform that logic in a second pass. So for my case I'm getting more then 2x performance boost. :)

> @jessegrosjean Sorry I didn't add this in my previous comment, but I'm super impressed you were able to get a working program while also being simultaneously new to Rust and working without any high level docs or examples. I've toiled quite a bit over the various APIs and worry quite a bit whether something is more complex than it needs to be. Abstraction has steep costs...

Ha, thanks. From my copy past and modify marathon I think the basic abstractions are pretty good. I had a whole lot of pain and confusion trying to figure out how rust in general worked, but the particular ripgrep API that I needed (minus the Bytes thing I guess) was pretty strait forward.

As far as ripgrep related code the part where I got most confused was with how the parallel walker worked. The actually documentation explains it pretty well. But I was just copy/pasting code and was very confused with this basic structure:

```
parallel_walker.run(move || {
        ...
        Box::new(move |result| {
            ...
        }
}
```

I didn't realize for a long time that the outer part was called only once per thread, and the inner part was called multiple times on that same thread.

The reason this wasn't obvious is because I didn't realize the outer part was returning the inner closure until long into the process. Especially hard I think when that closure has more then a few lines of logic.

It wasn't until I got things running in debugger that I figured out what was happening. And then to really finish things off I decided to read the docs, and that made it clear. I'm not sure you can do anything about it, seems like you are just following rust style, but an explicit `return Box::new(move |result| {` would have made things obvious and saved me a bunch of time I think.

**Comment by BurntSushi** [comment]
RE parallel walker: yes, I absolutely hate the API. Nested closures are just not good. I believe there's another issue tracking some ideas there, but yeah, that definitely needs rethinking but it probably won't be something I tackle for a while. I think in general the `ignore` crate needs a rethink, including its implementation. It is not easy code to maintain.

**Comment by NightMachinary** [comment]
Should I parse the json output now if I want to use ripgrep in rust? My main goal is customizing the output format so I can parse the cleaned result easily with `fzf`'s `--with-nth`.

**Comment by BurntSushi** [comment]
@NightMachinary I think Discussions would be a better place for questions like that: https://github.com/BurntSushi/ripgrep/discussions/1592
